### ISSUE TITLE
add parEval to run evaluations in parallel

### ISSUE THREAD

**Issue by Borgvall** [haskell/parallel#2]
Currently there are two functions inside the Eval monad two spark evaluations in parallel:

> rpar :: a -> Eval a -- rpar == rparWith rseq
> rparWith :: (a -> Eval a) -> a -> Eval a

Whats missing is a function

> parEval :: Eval a -> Eval a

with the semantic

> parEval . strat == rparWith strat -- edge case strat = r0

but works with all `Eval` values not just `Strategies`.

AFAIK there is no way to implement `parEval` using the public interface. I think it should be provided.

**Comment by Borgvall** [comment]
A candidate is

> parEval = rpar . runEval

but 
> rpar . runEval . r0 = rpar /= rparWith r0

**Comment by Borgvall** [comment]
I rethought about the problem, and it is possible to implement `parEval` with the public interface, but it looks odd:

> parEval eval = rparWith (const eval) undefined

Basically `parEval` is more primitive than `rparWith`, so it would be more appropriate to implement `rparWith` on top of `parEval`. I have done it in my fork (commits: https://github.com/Borgvall/parallel/commit/3aa862a3e016bf67f6c156eb36df1a804ef9a81b https://github.com/Borgvall/parallel/commit/3aca98d83a39a3ec8c18df4d8e25fbcbc19e5254 that depends on the older try to implement `parEval` https://github.com/Borgvall/parallel/commit/e387abaaa79e6034a6759c198454ad7ec62ddaa8). You might consider merging them.

**Comment by treeowl** [comment]
Your `parEvalC` looks right to me. ``rparWith s = rpar `dot` s``, and for all `t`, ``t `dot` r0 = t``. The latter follows from `runEval . r0 = id`.

**Comment by Borgvall** [comment]
The problem with `parEvalC` is, that `parEvalC (r0 x)` sparks the evaluation of `x` to weak head normal form, while `(rparWith r0) x` sparks an evaluation, that doesn't evaluate `x` at all. So `parEvalC` has not the semantic, I wanted for `parEval`.

**Comment by treeowl** [comment]
``rparWith r0 x = (rpar `dot` r0) x = rpar x``. `rparWith` just swallows `r0`, `rpar` and `rseq`.

**Comment by Borgvall** [comment]
`rparWith` is *not* `dot rpar` see Issue #35.

**Comment by treeowl** [comment]
Now that I am un-confused about `rparWith`, I see what you're getting at. It could be done like this:

```haskell
parEval m = do
  l <- rpar r
  return (case l of Lift x -> x)

  where
    r = runEval (Lift <$> m)
```

But I completely agree with you that `parEval` is more fundamental than `rparWith`, and that `rparWith` should indeed be built on top of it. Indeed, I think it would be pretty reasonable to argue that `parEval`, rather than `rpar`, should have been chosen as the primitive operation with native compiler support.

```haskell
rpar = parEval . rseq
```

**Comment by Borgvall** [comment]
When I opened this issue, I thought one needs the internal implementation of `Eval`, to implement `rparWith` or `parEval`. During conversation of issue #35, it turned out one can define both of them with the public interface, as @treeowl pointed out above.

I like the idea of having a primitive `parEval#`, but I don't know how big the impact in the RTS would be. One improvement would be, that the intermediate `Lift` box would vanish.

I can rebase, my old pull request #4 to master, but adding a primitive to GHC is out of scope for me.

**Comment by treeowl** [comment]
I don't think it's worth trying to add a primop now. A good spark will do way more work than the single extra object application demands. But we should definitely add `parEval` here.
